opm/simulators/wells/MultisegmentWell_impl.hpp

```diff
@@ -3681,26 +3697,39 @@ namespace Opm
         }
 
         EvalWell gas_fraction = 0.;
-        EvalWell gas_viscosities = 0.;
+        EvalWell gas_viscosity = 0.;
         if (FluidSystem::phaseIsActive(FluidSystem::gasPhaseIdx)) {
             const int gas_pos = Indices::canonicalToActiveComponentIndex(FluidSystem::gasCompIdx);
             gas_fraction = phase_fractions[gas_pos];
-            gas_viscosities = phase_viscosities[gas_pos];
+            gas_viscosity = phase_viscosities[gas_pos];
+        }
+
+        EvalWell density = segment_densities_[seg_upwind];
+        // WARNING
+        // We disregard the derivatives from the upwind density to make sure derivatives
+        // wrt. to different segments dont get mixed.
+        if (seg != seg_upwind) {
+            water_fraction.clearDerivatives();
+            water_viscosity.clearDerivatives();
+            oil_fraction.clearDerivatives();
+            oil_viscosity.clearDerivatives();
+            gas_fraction.clearDerivatives();
+            gas_viscosity.clearDerivatives();
+            density.clearDerivatives();
         }
 
         const EvalWell liquid_emulsion_viscosity = mswellhelpers::emulsionViscosity(water_fraction, water_viscosity,
                                                      oil_fraction, oil_viscosity, sicd);
-        const EvalWell mixture_viscosity = (water_fraction + oil_fraction) * liquid_emulsion_viscosity + gas_fraction * gas_viscosities;
+        const EvalWell mixture_viscosity = (water_fraction + oil_fraction) * liquid_emulsion_viscosity + gas_fraction * gas_viscosity;
 
-        const EvalWell& reservoir_rate = segment_reservoir_volume_rates_[seg];
```

Comment:
The code looks good to me. It is a sore point that we have not conclusively agreed on the correct mathematical model, but rather ended up with agreeing on a code that gives good results...

The only thing I'd like changed is: please remove completely the `segment_reservoir_volume_rates_` member, which is not used any more.

Reply:
ok